### DISCUSSION TITLE
Terminate on either WiFi or eth0 connection

### DISCUSSION THREAD

**Discussion by sistemicorp** [davesteele/comitup#149] [Q&A] [answered]
In my application, the Raspberry Pi (4B) can be connected either via WiFi or ethernet cable.

I would like the comitup hotspot to terminate if there is a connection on eth0.

Would there be some hints as to how to achieve this within the comitup code?

My application listens on port 80, so therefore it will not start when comitup hotspot is running.

I could have a shell script check for internet connection and then use `systemctl` to stop comitup.  

I am a python programmer so I could modify the code, but I am not a networking guy.  I would prefer to modify the comitup code so that if it found there was a connection on eth0, it would call the callback function with "CONNECTED".  I could use some hints about doing this.

**Comment by davesteele** [comment, marked as answer]
The place in the comitup code for that would be hotspot_start() and hotspot_timeout() in states.py.
... but that may not a particularly good place. Looking at the current code status, the timeout is sometimes called every 5 seconds, and sometimes every 3 minutes (if no AP connections are defined). I've done much fiddling with that number in the past - I'm not sure if setting it to 5 seconds always would be OK.
You could kill comitup externally, but you would also be responsible for tearing down the AP via nmcli.

**Comment by davesteele** [comment]
Logically, the comitup fix might be best solved by adding another state, for "ethernet active and kill-comitup-activity-when-ethernet is set".

**Reply by sistemicorp** [reply]
I see that you get the eth0 devices...

    def get_phys_dev_names():
        return [device_name(x) for x in get_devices() if x.DeviceType in (1, 2)]
 
Then make a `def get_eth_devices()`, and then check eth0 status...

Still researching...

**Reply by sistemicorp** [reply]
I think its fair to say, for my use case, that the wired ethernet connection would be in place before power-up.  So a check in `hotspot_start()`, or before would work.

**Reply by davesteele** [reply]
Unless the user plugs in the Ethernet cable later.

**Reply by davesteele** [reply]
Also note that the web service starts and stops based on callbacks called in the {state}.{start, pass, fail}() decorators.

**Reply by sistemicorp** [reply]
plugs in the Ethernet cable later: I can ignore this case for my application.  At least for a start, maybe down the road this can be improved.

I am slowly learning things in the code...

**Reply by sistemicorp** [reply]
Where are the known SSIDs/Passwords kept, I want the DUT to forget the network...
tried `/etc/wpa_supplicant`, `PERSIST_PATH`, `CONF_PATH`, `BOOT_CONF_PATH`, ...

**Reply by davesteele** [reply]
Managed by NetworkManager. You can do it by clearing the files in /etc/NetworkManager/system-connections if you are careful about it (NM may re-save on exit).

**Reply by sistemicorp** [reply]
I added a log here,

        def get_devices():
            global device_list
        
            if not device_list:
                try:
                    log.debug("Calling nm.GetDevices()")
                    device_list = nm.NetworkManager.GetDevices()
                except TypeError:
                    # NetworkManager is gone for some reason. Bail big time.
                    sys.exit(1)
        
            for d in device_list[1:]:
                if d.DeviceType == 1:
                    log.info("MGAG Type {}, ActiveConnection.State {}, Ip4 {}".format(d.DeviceType, 
                                                                                      d.ActiveConnection.State, 
                                                                                      d.Ip4Address))
        
            return device_list
            
and I can see this print, which tells me that wired ethernet is connected,

    comitup - INFO - MGAG Type 1, ActiveConnection.State 2, Ip4 192.168.0.131
    
But I am lost as to how to trigger `comitup` to shutdown...
Obviously where I put this log is during the initial setup of `comitup`... and presumably at this point there is no way to signal the shutdown...

Maybe adding the wired device to here,

        def init_nmmon():
            set_device_listeners(
                modemgr.get_ap_device(),
                modemgr.get_link_device()
            )

PS: is it okay to ask more questions?  Is this the right place to do it?  Perhaps slack or create a new issue?  I apologise if I am presuming you have the time or inclination to help.

**Reply by davesteele** [reply]
Questions are fine here. I reserve the right to stop answering at some point :-)

You are on the right track - nmmon.py is a good place to detect the ethernet port coming online. As for shutting down, the comitup service is set to auto-restart on exit. You may be able to shut down cleanly using sd_stop_unit in sysd.py.

If it were me, I'd make an external service which monitored the ethernet and killed comitup. Otherwise, you will be responsible to track any changes I make. If you do that, note that you are looking for a NetworkMonitor state of [NM_STATE_CONNECTED_GLOBAL](https://developer.gnome.org/NetworkManager/stable/nm-dbus-types.html#NMState), ethernet in NM_DEVICE_STATE_ACTIVATED,  and comitup in HOTSPOT mode.

The information is all available via D-Bus.

**Reply by sistemicorp** [reply]
I understand your point regarding external service and tracking your changes.

I think I am getting close, 

        2021-03-22 20:02:06,547 - comitup - INFO - No candidates - skipping CONNECTING scan
        2021-03-22 20:02:11,552 - comitup - INFO - No candidates - skipping CONNECTING scan
        2021-03-22 20:02:13,241 - comitup - INFO - MGAG: wired_changed_state 20
        2021-03-22 20:02:14,298 - comitup - INFO - MGAG: wired_changed_state 30
        2021-03-22 20:02:14,313 - comitup - INFO - MGAG: wired_changed_state 40
        2021-03-22 20:02:14,320 - comitup - INFO - MGAG: wired_changed_state 50
        2021-03-22 20:02:14,324 - comitup - INFO - MGAG: wired_changed_state 70
        2021-03-22 20:02:14,792 - comitup - INFO - MGAG: wired_changed_state 80
        2021-03-22 20:02:14,793 - comitup - INFO - MGAG nmm - primary pass, set_state CONNECTED
        2021-03-22 20:02:14,793 - comitup - INFO - MGAG state_matrix connected_start
        2021-03-22 20:02:14,793 - comitup - INFO - MGAG Setting state to CONNECTED, <function connected_start at 0xb5568618>
        2021-03-22 20:02:14,797 - comitup - INFO - MGAG: get_wired_device [<NetworkManager.Wired object at 0xb5579190>]
        2021-03-22 20:02:14,800 - comitup - INFO - MGAG state_matrix connected_pass
        2021-03-22 20:02:14,800 - comitup - INFO - MGAG state_matrix connected_fail
        2021-03-22 20:02:14,800 - comitup - INFO - MGAG state_matrix connected_timeout
        2021-03-22 20:02:14,801 - comitup - INFO - MGAG state_matrix connected_start
        2021-03-22 20:02:14,899 - comitup - INFO - MGAG external callback


In changed state callback, I do this,

        def wired_changed_state(state, *args):
            log.info("MGAG: wired_changed_state {}".format(state))
            if state in PASS_STATES:
                log.info("nmm - primary pass, set_state CONNECTED")
                set_state('CONNECTED')
            ...

But the Hostspot is still running and my app did not start (probably a listening port conflict).

**Reply by davesteele** [reply]
Regarding state callbacks, you should be aware of a bug. Some states call set_state directly, then exit. They should instead schedule execution as a callback. On exit, the callbacks are executed via a decorator. As a result, callbacks can be called out of order.

**Reply by sistemicorp** [reply]
I did get my initial test case working, that is on boot up, with eth0 plugged in, no WiFi available, it should start my application.  I added a flag that prevented the ap from accepting state changes once eth0 posted state CONNECTED.

The next problem I had is that I wasn't getting any events from wired eth0 when the cable was unplugged.   This is either a NetworkManager setup issue for that interface, or its a kernel bug that was supposed to be fixed in "5.10.y"... These issues are above my pay grade.

I am inclined to make my product "WiFi" only... and use comitup "as is".

If I may be so bold as to suggest another feature, for the hotspot SSID, it would be nice to have last 4 digits of MAC address, or RPi serial number.  So the config could be: <nn..> (existing random number) or <MMM...> or <SSS...> to indicate which one is used.  The idea is to have the user know which physical RPi (via label on box) is in hotspot mode.  Maybe this is something I can help prototype if there is interest.

**Reply by davesteele** [reply]
any_changed_state() in nmmon.py should be getting called for eth0 state changes.

I try to give reasonably quick reviews to PRs.

**Reply by sistemicorp** [reply]
yeah, I can't understand why I see state changes for connecting, but none for disconnecting.  Maybe this,
https://github.com/raspberrypi/firmware/issues/1100

I did a test where I logged `nmcli m` output into a file, and I pulled the cable in/out... and as expected,

    'p1125-451-0000' is now the primary connection
    'Wired connection 1' is now the primary connection

And I think this verifies that in this state, comitup, is still running,

    $ ps -ef | grep comitup
    root       501     1  0 17:33 ?        00:00:01 /usr/bin/python3 /usr/sbin/comitup

**Reply by davesteele** [reply]
> If I may be so bold as to suggest another feature, for the hotspot SSID, it would be nice to have last 4 digits of MAC address, or RPi serial number

I'd be happy to see those options show up for [_ap\_name_](https://davesteele.github.io/comitup/man/comitup-conf.5.html).

**Reply by sistemicorp** [reply]
My intent is to start today.  I already have python code that can get MAC and RPi Serial number from another project I have. 

Any suggestions for why I don't see `eth0` disconnecting event(s)?
My only lead is that there perhaps should be an entry in `NetworkManager.conf` for eth0.... But I think `comitup` manages that file, so I was reluctant to touch it yet, without further understanding.  

I also wondering if there is a way to "kick" NetworkManager to make it publish all the current states of the devices (though specifically `eth0`).  I think I need this because on boot, almost half the time `comitup` doesn't get the `eth0` connecting states, I assume those events happened before `comitup` service was running (as the `eth0` cable was plugged in on power up).  Googling the APIs I haven't seen the way to do this.  The other way would be to read `eth0` state when `comitup` initialises...

**Reply by davesteele** [reply]
> Any suggestions for why I don't see `eth0` disconnecting event(s)?

No. any_changed_state() should be seeing the state changes associated with that. There may be another NM signal to listen to.

**Reply by davesteele** [reply]
> an entry in `NetworkManager.conf` for eth0.... But I think `comitup` manages that file

No. The Comitup Image defines a pair of NM connections, so that it will auto-connect DHCP or static, as appropriate. But it is all NM after that, and the package doesn't know from wired connections.

**Reply by sistemicorp** [reply]
I have working prototype for specifying the use of partial MAC or RPi serial number, in the same way that random number was used.  Before I post code, a few comments to clarify things, and/or to change the way it works...
1) The MAC and SN use the back end digits, not from the front.  For example if the serial number is `10000123`, then specifying `<sss>` will get you `123`.
2) I threw in extra code that determines if the MAC or SN needs to be fetched, if `ap_name` doesn't call for it, then don't bother getting it.  This is an attempt to have no impact on existing (working) systems.  The new MAC/SN fetching code might not work on an older system or something.  I am always paranoid. 
3) I added some validation... if the `ap_name` spec in the conf file doesn't look right, fall back to default.
